crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -198,50 +202,38 @@ pub(crate) fn impl_enum(reflect_enum: &ReflectEnum) -> proc_macro2::TokenStream
                 #FQOption::Some(<Self as #bevy_reflect_path::Typed>::type_info())
             }
 
-            #[inline]
-            fn into_any(self: #FQBox<Self>) -> #FQBox<dyn #FQAny> {
-                self
+            fn try_as_reflect(&self) -> #FQOption<&dyn #bevy_reflect_path::Reflect> {
+                #FQOption::Some(self)
             }
 
-            #[inline]
-            fn as_any(&self) -> &dyn #FQAny {
-                self
+            fn try_as_reflect_mut(&mut self) -> #FQOption<&mut dyn #bevy_reflect_path::Reflect> {
+                #FQOption::Some(self)
             }
 
-            #[inline]
-            fn as_any_mut(&mut self) -> &mut dyn #FQAny {
-                self
+            fn try_into_reflect(self: Box<Self>) -> #FQResult<Box<dyn #bevy_reflect_path::Reflect>, Box<dyn #bevy_reflect_path::PartialReflect>> {
+                #FQResult::Ok(self)
             }
 
-            #[inline]
-            fn into_reflect(self: #FQBox<Self>) -> #FQBox<dyn #bevy_reflect_path::Reflect> {
+            fn as_partial_reflect(&self) -> &dyn #bevy_reflect_path::PartialReflect {
                 self
             }
 
-            #[inline]
-            fn as_reflect(&self) -> &dyn #bevy_reflect_path::Reflect {
+            fn as_partial_reflect_mut(&mut self) -> &mut dyn #bevy_reflect_path::PartialReflect {
                 self
             }
 
-            #[inline]
-            fn as_reflect_mut(&mut self) -> &mut dyn #bevy_reflect_path::Reflect {
+            fn into_partial_reflect(self: #FQBox<Self>) -> #FQBox<dyn #bevy_reflect_path::PartialReflect> {
```

Comment:
Does `[as,into]_partial_reflect[_mut]` make sense? Maybe it makes sense to remove them. Since anything that implements `Reflect` also implements `PartialReflect`, and it's rare to have a `T: !Reflect` outside of `dyn PartialReflect`.

Reply:
these are crucial for the following situation:
```rust
fn partial_introspection(value: &dyn PartialReflect);

// There are situations where concrete `T: Reflect` is not applicable enough
// (e.g. working with type data and the registry).
fn concrete_introspection(value: &dyn Reflect) {
    // there's no other existing way to do this conversion:
    partial_introspection(value.as_partial_reflect());
}
```

Reply:
I see. Rust do need trait object downcasting for this to work without the helper method